cheatsheets/Forgot_Password_Cheat_Sheet.md

```diff
@@ -1,60 +1,137 @@
 
-## Step 4) Allow user to change password in the existing session
+## Methods
 
-Step 4 requires input of the code sent in step 3 in the existing session where the challenge questions were answered in step 2, and allows the user to reset his password. Display a simple HTML form with one input field for the code, one for the new password, and one to confirm the new password. Verify the correct code is provided and be sure to enforce all password complexity requirements that exist in other areas of the application. 
+In order to allow a user to request a password reset, you will need to have some way to identify the user, or a means to reach out to them through a side-channel.
 
-As before, avoid sending the username as a parameter when the form is submitted. Finally, it's critical to have a check to prevent a user from accessing this last step without first completing steps 1 and 2 correctly. Otherwise, a [forced browsing](https://owasp.org/www-community/attacks/Forced_browsing) attack may be possible. Ensure the user changes their password and does not simply surf to another page in the application. 
+This can be done through any of the following methods:
 
-The reset must be performed before any other operations can be performed by the user.
+- One Time Password (OTP).
+- URL tokens.
+- Security Questions.
+- Offline backup codes.
 
-## Step 5) Logging
+> These methods can be used together and many times it is recommended to do so. No matter what you must ensure that a user always has a way to recover their account.
 
-It is important to keep audit records when password change requests were submitted. This includes whether or not security questions were answered, when reset messages were sent to users and when users utilize them. It is especially important to log failed attempts to answer security questions and failed attempted use of expired tokens. This data can be used to detect abuse and malicious behavior. Data such as time, IP address, and browser information can be used to spot trends of suspicious use.
+## Tokens and Codes Secure Practices
 
-# Other Considerations
+It is essential to employ security practices for the reset codes and tokens that will be used in the methods.
 
-- Whenever a successful password reset occurs, all other sessions should be invalidated. Note the current session is already authenticated and does not require a login prompt.
-- Strength of questions used for reset should vary based on the nature of the credential. Administrator credentials should have a higher requirement.
-- The ideal implementation should rotate the questions asked in order to avoid automation.
+**Common secure practices**:
+
+- Randomly generated with a [CSPRNG](https://en.wikipedia.org/wiki/Cryptographically_secure_pseudorandom_number_generator) (*e.g.* [secrets](https://docs.python.org/3/library/secrets.html) library in Python).
```

Comment:
We have a section on secure randomness in another CS (Cryptographic storage?). Link to that